wasm/validation.cpp

```diff
@@ -0,0 +1,520 @@
+struct ControlFrame {
+    Instruction i;
+    std::vector<ValueType> params;
+    std::vector<ValueType> results;
+    std::size_t stack_height = 0;
+    bool unreachable = false;
+};
+
+struct Validator {
+    std::vector<ValueOrUnknown> value_stack; // operand stack
+    std::vector<ControlFrame> control_stack;
+
+    void push_val(ValueOrUnknown const &);
+    tl::expected<ValueOrUnknown, ValidationError> pop_val();
+    tl::expected<ValueOrUnknown, ValidationError> pop_val_expect(ValueOrUnknown const &);
+    void push_vals(std::vector<ValueType> const &);
+    tl::expected<std::vector<ValueOrUnknown>, ValidationError> pop_vals(std::vector<ValueType> const &);
+    void push_ctrl(Instruction const &, std::vector<ValueType> const &, std::vector<ValueType> const &);
+    tl::expected<ControlFrame, ValidationError> pop_ctrl();
+    std::vector<ValueType> const &label_types(ControlFrame &);
+    void unreachable();
+};
+
+void Validator::push_val(ValueOrUnknown const &val) {
+    value_stack.push_back(val);
+}
+
+tl::expected<ValueOrUnknown, ValidationError> Validator::pop_val() {
+    assert(!control_stack.empty());
+
+    if (value_stack.size() == control_stack.back().stack_height && control_stack.back().unreachable) {
+        return Unknown{};
+    }
+
+    if (value_stack.size() == control_stack.back().stack_height) {
+        return tl::unexpected{ValidationError::ValueStackUnderflow};
+    }
+
+    auto val = value_stack.back();
+
+    value_stack.pop_back();
+
+    return val;
+}
+
+tl::expected<ValueOrUnknown, ValidationError> Validator::pop_val_expect(ValueOrUnknown const &expected) {
+    auto actual = pop_val();
+
+    if (!actual.has_value()) {
+        return tl::unexpected{actual.error()};
+    }
+
+    if (actual != expected && !std::holds_alternative<Unknown>(*actual) && !std::holds_alternative<Unknown>(expected)) {
+        return tl::unexpected{ValidationError::ValueStackUnexpected};
+    }
+
+    return actual;
+}
+
+void Validator::push_vals(std::vector<ValueType> const &vals) {
+    value_stack.reserve(value_stack.size() + vals.size());
+
+    for (ValueType const &val : vals) {
+        value_stack.push_back(val);
+    }
+}
+
+tl::expected<std::vector<ValueOrUnknown>, ValidationError> Validator::pop_vals(std::vector<ValueType> const &vals) {
+    std::vector<ValueOrUnknown> popped;
+
+    // TODO(dzero): switch to std::ranges::reverse_view once we drop older toolchains
+    // NOLINTNEXTLINE(modernize-loop-convert)
+    for (auto it = vals.rbegin(); it != vals.rend(); it++) {
+        auto maybe_val = pop_val_expect(*it);
+
+        if (!maybe_val.has_value()) {
+            return tl::unexpected{maybe_val.error()};
+        }
+
+        popped.insert(popped.begin(), *maybe_val);
+    }
+
+    return popped;
+}
+
+void Validator::push_ctrl(
+        Instruction const &i, std::vector<ValueType> const &params, std::vector<ValueType> const &results) {
```

Comment:
These are more suitable to pass by value and move into the `ControlFrame` since they're either heap-allocated or can contain heap-allocated objects.